pkg/features/kube_features.go

```diff
@@ -970,7 +969,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	DynamicResourceAllocation: {Default: false, PreRelease: featuregate.Alpha},
 
-	EventedPLEG: {Default: false, PreRelease: featuregate.Beta}, // off by default, requires CRI Runtime support
+	EventedPLEG: {Default: false, PreRelease: featuregate.Alpha},
```

Comment:
During the discussion on Slack I was assuming that this must have been a change that went recently into master for 1.30 and some real bug.

However, now that I look at this I'm coming to a slightly different conclusion: as the original comment says, this is not a feature that can be enabled in all environments. That comment got lost when promoting to beta.

I think this is a misuse of the feature gate mechanism. If this cannot be enabled unconditionally, then there should be a separate kubelet (?) command line flag or configuration option for enabling it.

As it stands, this can only be promoted to GA (which implies removing the feature gate eventually!) if *all* CRI runtimes support this. Is that the goal? I haven't checked the KEP.

Reply:
for one or other reason, this is clear no beta quality, I just found also that this issue was already reported in alpha an never addresses https://github.com/kubernetes/kubernetes/issues/121003

Reply:
Do we have a precedence for downgrading a feature after it was released?

I agree that it should have remained alpha. I just want to understand whether downgrading really is an option.

Reply:
I removed one entirely some time ago https://github.com/kubernetes/kubernetes/pull/121229

I don't know about precedence and this is better to consult as you say, @liggitt do you have any memories?

Reply:
>I think this is a misuse of the feature gate mechanism. If this cannot be enabled unconditionally, then there should be a separate kubelet (?) command line flag or configuration option for enabling it.

Evented PLEG requires a CRI runtime that supports generating those events. [KEP](https://github.com/kubernetes/enhancements/blob/5895f7d5a0506b03266704b82ecd551ff180925b/keps/sig-node/3386-kubelet-evented-pleg/README.md#compatibility-check) talks about falling back on Generic PLEG in case the kubelet with Evented PLEG is paired with the runtime that doesn't support generating the required events. So ideally, this feature should have worked unconditionally whether a CRI runtime supports Evented PLEG or not. 

This has been [handled](https://github.com/kubernetes/kubernetes/blob/773ad73b90c2836685358a0fafe285b9e4cdcc19/pkg/kubelet/pleg/evented.go#L185) in the code.

Reply:
So removing that comment was intentional - so far, so good.

But the observation remains that if this feature gate is enabled, testing fails. https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind-beta-features passes with this PR (because the job does not enable alpha features) and fails without it (because it turns on EventedPLEG).

If you have some other PR that makes pull-kubernetes-e2e-kind-beta-features then this might be an alternative to reverting.

Reply:
> Do we have a precedence for downgrading a feature after it was released?

yes, when a feature was determined not to meet graduation requirements and cause regressions

> ConfigMap rendering [issue](https://github.com/kubernetes/kubernetes/issues/112081) was found in the 1.25.0 release. When ConfigMaps get updated within the API, they do not get rendered to the resulting pod's filesystem by the Kubelet. The feature has been [reverted to alpha](https://github.com/kubernetes/kubernetes/pull/112078) in the 1.25.1 release.

Reply:
whatever change we make here should be consistently backported to 1.27/1.28/1.29

since the feature was never enabled by default, changing the state will not impact users who are currently explicitly enabling and using it (they can still do that if they want), but does accurately convey the stability level

Reply:
Although from the [slack conversation](https://kubernetes.slack.com/archives/C09QZ4DQB/p1704956800671419?thread_ts=1704859442.079249&cid=C09QZ4DQB) it seems the Evented PLEG feature is completely disabled then the beta job passes.

Reply:
The job passes when it doesn't enable Evented PLEG. It should pass also when it is enabled because it should be safe to enable all beta features.

I'm not sure whether https://github.com/kubernetes/kubernetes/pull/122124 is a proper fix for this problem. I'll let people more familiar with the feature discuss that in that PR.

Even if that is fixed, is the quality and test coverage of Evented PLEG really good enough to call it beta?

The usual graduation criteria for beta are ["e2e tests completed and enabled"](https://github.com/kubernetes/enhancements/blob/5895f7d5a0506b03266704b82ecd551ff180925b/keps/NNNN-kep-template/README.md?plain=1#L369) and for beta, [links are expected](https://github.com/kubernetes/enhancements/blob/5895f7d5a0506b03266704b82ecd551ff180925b/keps/NNNN-kep-template/README.md?plain=1#L329-L335).

It looks like for PLEG, using the feature in a "real" cluster wasn't considered and not tested (https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3386-kubelet-evented-pleg#graduation-criteria only mentions E2E node tests). E2E node testing is not sufficient.

Given that there were undetected problems, I'm :+1: for downgrading to alpha, updating the KEP with a more complete test plan, then upgrading to beta when known issues are fixed and test coverage is more complete.

/cc @aojea